feat: support query-based requests

itsHenry35 opened the pull request:
fix https://github.com/rclone/gofakes3/issues/5
fix https://github.com/rclone/rclone/issues/7616

KevinWang15 commented:
Hi @itsHenry35,

Thank you for your timely pull request.

I've tested your code, and I noticed that [this change](https://github.com/rclone/gofakes3/pull/6/files#diff-5306d5b0f8c88b148ffa772f859fed982963272c2fe89c9f1e18b544075afdcfL160-R174) might not always be correct.

<img src="https://github.com/rclone/gofakes3/assets/10173472/25b21d37-1f74-494e-8020-6e8716e68833" width="200"/>

In most cases, `amzDate` is present in the header.

As an alternative, we could consider the following approach:

```go
var date string
if date = req.Header.Get(amzDate); date == "" {
    if date = queryf.Get(amzDate); date == "" {
        if date = r.Header.Get(headerDate); date == "" {
            return errMissingDateHeader
        }
    }
}
```

Additionally, I'd like to suggest squashing your commits to remove the "[chore: remove generated binary](https://github.com/rclone/gofakes3/pull/6/commits/c2cadeb0f5db83d04d75701c3cb6696ec7f47c6e)" commit before submitting this PR (or you could use `git commit --amend`).

Once again, thanks for the excellent work!

itsHenry35 commented:
> Hi @itsHenry35,
> 
> Thank you for your timely pull request.
> 
> I've tested your code, and I noticed that [this change](https://github.com/rclone/gofakes3/pull/6/files#diff-5306d5b0f8c88b148ffa772f859fed982963272c2fe89c9f1e18b544075afdcfL160-R174) might not always be correct.
> 
> <img alt="" width="200" src="https://private-user-images.githubusercontent.com/10173472/346310166-25b21d37-1f74-494e-8020-6e8716e68833.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAzMjU3MjcsIm5iZiI6MTcyMDMyNTQyNywicGF0aCI6Ii8xMDE3MzQ3Mi8zNDYzMTAxNjYtMjViMjFkMzctMWY3NC00OTRlLTgwMjAtNmU4NzE2ZTY4ODMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA3VDA0MTAyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIxYzc2MTMyZDZiMTM0NWZhMTI2ZWJhNmNiZjUzZjA0NzNkODU0NzE3MTY0MjE4NGMyZDAyOTAwNmEzOTZjOTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QC8D1-GAsUaWtrl5IPdVEyrAk1lLhegcuysPcTzqK2o">
> In most cases, `amzDate` is present in the header.
> 
> As an alternative, we could consider the following approach:
> 
> ```go
> var date string
> if date = req.Header.Get(amzDate); date == "" {
>     if date = r.Header.Get(headerDate); date == "" {
>         if date = queryf.Get(amzDate); date == "" {
>             return errMissingDateHeader
>         }
>     }
> }
> ```
> 
> Additionally, I'd like to suggest squashing your commits to remove the "[chore: remove generated binary](https://github.com/rclone/gofakes3/pull/6/commits/c2cadeb0f5db83d04d75701c3cb6696ec7f47c6e)" commit before submitting this PR (or you could use `git commit --amend`).
> 
> Once again, thanks for the excellent work!

Oh, i was opposed to just add a if judgement instead of replacing it, but I did it by mistake. Sorry for the fault, I will correct it

itsHenry35 committed:
feat(auth): enhance AWS Signature V4 authentication support

- Add query-based authentication alongside header-based
- Support "UNSIGNED-PAYLOAD"
- Add fallback options for date header extraction
- Fix ordering of query keys by using req.URL.Query().Encode() instead of req.URL.RawQuery

KevinWang15 commented:
@itsHenry35 Great, I have tested your code. Now it works. LGTM

cc @ncw PTAL

ncw commented on signature/signature-v4.go:
```diff
@@ -134,9 +140,17 @@ func V4SignVerify(r *http.Request) ErrorCode {
+	if v4Auth == "" {
+		// try query based
+		v4Auth = fmt.Sprintf("%s Credential=%s,SignedHeaders=%s, Signature=%s", queryf.Get(amzAlgorithm), queryf.Get(amzCredential), queryf.Get(amzSignedHeaders), queryf.Get(amzSignature))
+		if v4Auth == "" {
```
This will never be true. What is this test for?

itsHenry35 replied:
https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
Please have a look as a reference. When the client generates a web url for easier accessing on web browsers, the authorization headers are passed in query-mode instead of headers. In this case, the v4Auth will be "" and we can get the  auth params from the url query.

KevinWang15 replied:
This code change is for support for "pre-signed URLs", a feature quite commonly used in S3.

Example (https://runkit.com/kevinwang15/668e1f70270f8e0008046ad9):

```javascript
const AWS = require("aws-sdk");

const s3 = new AWS.S3({
    endpoint: 'https://127.0.0.1',
    accessKeyId: 'aaaa',
    secretAccessKey: 'ssss',
    s3ForcePathStyle: true,
    signatureVersion: 'v4',
    sslEnabled: true,
});


function generateDownloadUrl(bucket, key) {
    const params = {
        Bucket: bucket,
        Key: key,
        Expires: 3600,
    };

    s3.getSignedUrl('getObject', params, (err, url) => {
        if (err) {
            console.error(err);
        } else {
            console.log(url);
        }
    });
}

generateDownloadUrl('bucket1', 'a.txt');

```

It produces URLs like

> https://127.0.0.1/bucket1/a.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=aaaa%2F20240710%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240710T053620Z&X-Amz-Expires=3600&X-Amz-Signature=623b5d21aa2698af14d046d40008a72ef4ba3dfed9bb90d412e6dba4a6ec8706&X-Amz-SignedHeaders=host

This URL can be used anywhere - send a GET request and it will give you the file.

ncw replied:
I think we are talking at cross purposes! I agree with what you said above but it wasn't my question.

My specific query is this:

In this line we set `v4Auth` to something

```go
v4Auth = fmt.Sprintf("%s Credential=%s, SignedHeaders=%s, Signature=%s", queryf.Get(amzAlgorithm), queryf.Get(amzCredential), queryf.Get(amzSignedHeaders), queryf.Get(amzSignature))
```

This can never be `""`

And in the line after we check it against `""` which can never be true surely?

```go
if v4Auth == "" {
```

Which is indicative of some kind of logic error.

itsHenry35 replied:
Uh sorry, it's my fault, I'm going to correct it.

itsHenry35 replied:
Thanks for correction. I've corrected this logical error, pls have a look at the new force-pushed code.

ncw commented on signature/signature-v4.go:
```diff
@@ -159,7 +173,9 @@ func V4SignVerify(r *http.Request) ErrorCode {
 	if date = req.Header.Get(amzDate); date == "" {
 		if date = r.Header.Get(headerDate); date == "" {
-			return errMissingDateHeader
+			if date = queryf.Get(amzDate); date == "" {
```
I'd prefer these on two lines as at first glance it appears you've shadowed `date` using a very common go idiom but you haven't as there is an `=` instead of a `:=`

```go
date = queryf.Get(amzDate)
if date == "" {
```

ncw commented on signature/signature-v4.go:
```diff
@@ -186,7 +203,14 @@ func V4SignVerify(r *http.Request) ErrorCode {
 	// Verify if signature match.
 	if !compareSignatureV4(newSignature, signV4Values.Signature) {
-		return errSignatureDoesNotMatch
+		//try unsigned payload
```
Can't we detect the user is using "UNSIGNED-PAYLOAD" by looking in `X-Amz-Content-Sha256`? Then we don't need to guess .

itsHenry35 replied:
Thanks for the suggestion. That sounds reasonable, I'll do some further tests.

ncw reviewed (commented):
Thank you for doing this - every step towards greater compatibility is a good one :-)

I put some comments inline for discussion.

Thanks

Nick

**Edit** we should also think of some tests!

KevinWang15 commented:
Code has been updated, LGTM
@ncw PTAL again, thanks

ncw reviewed (approved):
This is looking great now - thank you :-)

I'll just wait for the CI to go green then I'll merge.

Do you want to propose a PR for rclone to update the go.mod to fix the corresponding rclone issue?

itsHenry35 commented:
> This is looking great now - thank you :-)
> 
> I'll just wait for the CI to go green then I'll merge.
> 
> Do you want to propose a PR for rclone to update the go.mod to fix the corresponding rclone issue?

Yeah, of course. I will propose the PR after the merge. Thx ^:^

ncw commented:
Thank you - great job :-)

itsHenry35 commented:
> This is looking great now - thank you :-)
> 
> I'll just wait for the CI to go green then I'll merge.
> 
> Do you want to propose a PR for rclone to update the go.mod to fix the corresponding rclone issue?

@ncw I've submitted the PR. Feel free to review it. https://github.com/rclone/rclone/pull/7944
